### DISCUSSION TITLE
What is a NullPointerException, and how do I fix it?

### DISCUSSION THREAD

**Discussion by toshikijp** [andatoshiki/toshiki-discuss#26] [Q&A] [answered]
What are Null Pointer Exceptions `java.lang.NullPointerException` and what causes them?

What methods/tools can be used to determine the cause so that you stop the exception from causing the program to terminate prematurely?

**Comment by andatoshiki** [comment, marked as answer]
When you declare a reference variable (i.e., an object), you are really creating a pointer to an object. Consider the following code where you declare a variable of primitive type `int`:

```java
int x;
x = 10;
```

In this example, the variable `x` is an `int` and Java will initialize it to `0` for you. When you assign the value of `10` on the second line, your value of `10` is written into the memory location referred to by `x`.

But, when you try to declare a reference *type*, something different happens. Take the following code:

```java
Integer num;
num = new Integer(10);
```

The first line declares a variable named `num`, but it does not actually contain a primitive value yet. Instead, it contains a pointer (because the type is `Integer` which is a reference type). Since you have not yet said what to point to, Java sets it to `null`, which means "**I am pointing to \*nothing\***".

In the second line, the `new` keyword is used to instantiate (or create) an object of type `Integer`, and the pointer variable `num` is assigned to that `Integer` object.

The `NullPointerException` (NPE) occurs when you declare a variable but did not create an object and assign it to the variable before trying to use the contents of the variable (called *dereferencing*). So you are pointing to something that does not actually exist.

Dereferencing usually happens when using `.` to access a method or field, or using `[` to index an array.

If you attempt to dereference `num` *before* creating the object you get a `NullPointerException`. In the most trivial cases, the compiler will catch the problem and let you know that "`num may not have been initialized`," but sometimes you may write code that does not directly create the object.

For instance, you may have a method as follows:

```java
public void doSomething(SomeObject obj) {
   // Do something to obj, assumes obj is not null
   obj.myMethod();
}
```

In which case, you are not creating the object `obj`, but rather assuming that it was created before the `doSomething()` method was called. Note, it is possible to call the method like this:

```java
doSomething(null);
```

In which case, `obj` is `null`, and the statement `obj.myMethod()` will throw a `NullPointerException`.

If the method is intended to do something to the passed-in object as the above method does, it is appropriate to throw the `NullPointerException` because it's a programmer error and the programmer will need that information for debugging purposes.

In addition to `NullPointerException`s thrown as a result of the method's logic, you can also check the method arguments for `null` values and throw NPEs explicitly by adding something like the following near the beginning of a method:

```java
// Throws an NPE with a custom error message if obj is null
Objects.requireNonNull(obj, "obj must not be null");
```

Note that it's helpful to say in your error message clearly *which* object cannot be `null`. The advantage of validating this is that 1) you can return your own clearer error messages and 2) for the rest of the method you know that unless `obj` is reassigned, it is not null and can be dereferenced safely.

Alternatively, there may be cases where the purpose of the method is not solely to operate on the passed in object, and therefore a null parameter may be acceptable. In this case, you would need to check for a **null parameter** and behave differently. You should also explain this in the documentation. For example, `doSomething()` could be written as:

```java
/**
  * @param obj An optional foo for ____. May be null, in which case
  *  the result will be ____.
  */
public void doSomething(SomeObject obj) {
    if(obj == null) {
       // Do something
    } else {
       // Do something else
    }
}
```

Finally, [[How to pinpoint the exception & cause using Stack Trace](https://stackoverflow.com/q/3988788/2775450)](https://stackoverflow.com/q/3988788/2775450)

> What methods/tools can be used to determine the cause so that you stop the exception from causing the program to terminate prematurely?

Sonar with find bugs can detect NPE. [[Can sonar catch null pointer exceptions caused by JVM Dynamically](https://stackoverflow.com/questions/20899931/can-sonar-catch-null-pointer-exceptions-caused-by-jvm-dynamically)](https://stackoverflow.com/questions/20899931/can-sonar-catch-null-pointer-exceptions-caused-by-jvm-dynamically)

Now Java 14 has added a new language feature to show the root cause of NullPointerException. This language feature has been part of SAP commercial JVM since 2006.

In Java 14, the following is a sample NullPointerException Exception message:

> in thread "main" java.lang.NullPointerException: Cannot invoke "java.util.List.size()" because "list" is null
